### PR TITLE
Feature/48 handling response status codes

### PR DESCRIPTION
**Pull request by arekwolk**
for issue https://github.com/Cognifide/knotx/issues/48

---

I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: knotx-core/knotx-api/src/main/java/com/cognifide/knotx/api/ErrorRepositoryResponse.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ * limitations under the License.
+ */
+package com.cognifide.knotx.api;
+
+import io.vertx.rxjava.core.MultiMap;
+
+public class ErrorRepositoryResponse extends RepositoryResponse {
+    public ErrorRepositoryResponse(int statusCode, String reason, MultiMap headers) {
+        super.statusCode = statusCode;
+        super.reason = reason;
```
Comment (marcinczeczko) [inline knotx-core/knotx-api/src/main/java/com/cognifide/knotx/api/ErrorRepositoryResponse.java]: Not a reason. It must be content - body of response

File: knotx-core/knotx-api/src/main/java/com/cognifide/knotx/api/EmptyRepositoryResponse.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.cognifide.knotx.api;
+
+import io.vertx.rxjava.core.MultiMap;
+
+public class EmptyRepositoryResponse extends RepositoryResponse {
```
Comment (tomaszmichalak) [inline knotx-core/knotx-api/src/main/java/com/cognifide/knotx/api/EmptyRepositoryResponse.java]: We should provide 2 additional subclasses of RepositoryResponse:
DynamicRepositoryResponse (Repository knows that requested response contains dynamic snippets)
StaticRepositoryResponse (Repository knows that requested response does not contain dynamic snippets)

File: knotx-core/knotx-api/src/main/java/com/cognifide/knotx/api/RepositoryResponse.java [hunk 3]
```diff
@@ -3,84 +3,70 @@
...
+        MultiMap headers = fromJsonArray(object.getJsonArray("headers"));
+        Boolean shouldProcess = object.getBoolean("shouldProcess");
+        int statusCode = object.getInteger("statusCode");
 
-    public RepositoryResponse(JsonObject object) {
-        success = object.getBoolean("success");
-        headers = fromJsonArray(object.getJsonArray("headers"));
-        if (success) {
-            data = object.getString("data");
+        if (statusCode == 200 && shouldProcess) {
```
Comment (tomaszmichalak) [inline knotx-core/knotx-api/src/main/java/com/cognifide/knotx/api/RepositoryResponse.java]: I don't like this if statement. In current implementation I am wondering if we really need EmptyRepositoryResponse (it is not SuccessRepositoryResponse)? It is more like special kind of it. Then we can remove this if case: "statusCode == 200 && shouldProcess".

File: knotx-core/knotx-repository/src/main/java/com/cognifide/knotx/repository/impl/HttpRepository.java [hunk 4]
```diff
@@ -74,16 +75,21 @@ static HttpRepository of(String path, String domain, Integer port, Vertx vertx)
...
                 .mergeWith(response.toObservable())
                 .reduce(Buffer::appendBuffer)
-                .flatMap(buffer -> toRepositoryResponse(buffer,response));
+                .flatMap(buffer -> toRepositoryResponse(buffer, response));
     }
 
     private Observable<RepositoryResponse> toRepositoryResponse(Buffer buffer, final HttpClientResponse httpClientResponse) {
         Observable<RepositoryResponse> response;
-        if (buffer.length() > 0) {
+        if (httpClientResponse.statusCode() == 200 && buffer.length() > 0) {
```
Comment (tomaszmichalak) [inline knotx-core/knotx-repository/src/main/java/com/cognifide/knotx/repository/impl/HttpRepository.java]: The same as previous.

File: knotx-core/knotx-repository/src/main/java/com/cognifide/knotx/repository/impl/HttpRepository.java [hunk 5]
```diff
@@ -74,16 +75,21 @@ static HttpRepository of(String path, String domain, Integer port, Vertx vertx)
...
             response = RepositoryResponse.success(buffer.toString(), httpClientResponse.headers()).toObservable();
+        } else if (httpClientResponse.statusCode() == 200) {
+            response = RepositoryResponse.empty(httpClientResponse.headers()).toObservable();
         } else {
-            LOGGER.info("Remote repository returned empty template for path `{}` with status code {}", path, httpClientResponse.statusCode());
-            response = RepositoryResponse.error(httpClientResponse.statusCode(), httpClientResponse.headers()).toObservable();
+            LOGGER.info("Remote repository returned empty template for path `{}` with status code {}", path,
```
Comment (marcinczeczko) [inline knotx-core/knotx-repository/src/main/java/com/cognifide/knotx/repository/impl/HttpRepository.java]: Wrong message to logger. Check the logic it's not about emptiness

Comment (arekwolk) [inline knotx-core/knotx-repository/src/main/java/com/cognifide/knotx/repository/impl/HttpRepository.java]: fixed

File: knotx-core/knotx-api/src/main/java/com/cognifide/knotx/api/ErrorRepositoryResponse.java (hunk 1 above)
Comment (arekwolk) [inline knotx-core/knotx-api/src/main/java/com/cognifide/knotx/api/ErrorRepositoryResponse.java]: fixed

**Comment by Skejven** [comment]
Please add proper pull request message according to [Contributing Rules](https://github.com/Cognifide/knotx/blob/master/CONTRIBUTING.md#pull-requests).

File: knotx-core/knotx-api/src/main/java/com/cognifide/knotx/api/RepositoryResponse.java [hunk 6]
```diff
@@ -3,92 +3,65 @@
  *
  * Copyright (C) 2016 Cognifide Limited
  *
- * Licensed under the Apache License, Version 2.0 (the "License");
- * you may not use this file except in compliance with the License.
- * You may obtain a copy of the License at
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in
```
Comment (Skejven) [inline knotx-core/knotx-api/src/main/java/com/cognifide/knotx/api/RepositoryResponse.java]: Please don't format this header.

File: knotx-core/knotx-api/src/main/java/com/cognifide/knotx/api/RepositoryResponse.java [hunk 7]
```diff
@@ -3,92 +3,65 @@
...
 package com.cognifide.knotx.api;
 
 import io.vertx.core.json.JsonObject;
 import io.vertx.rxjava.core.MultiMap;
 import rx.Observable;
 
-public class RepositoryResponse extends JsonObjectRequest {
+public abstract class RepositoryResponse extends JsonObjectRequest {
 
     private static final long serialVersionUID = -5892594714118271978L;
```
Comment (Skejven) [inline knotx-core/knotx-api/src/main/java/com/cognifide/knotx/api/RepositoryResponse.java]: Please update `serialVersionUID` after the class changed.

File: knotx-core/knotx-api/src/main/java/com/cognifide/knotx/api/RepositoryResponse.java [hunk 8]
```diff
@@ -3,92 +3,65 @@
...
-    private RepositoryResponse() {
-        //Empty default constructor
-    }
-
-    public RepositoryResponse(JsonObject object) {
-        success = object.getBoolean("success");
-        headers = fromJsonArray(object.getJsonArray("headers"));
-        if (success) {
-            data = object.getString("data");
+        if (statusCode == 200) {
```
Comment (Skejven) [inline knotx-core/knotx-api/src/main/java/com/cognifide/knotx/api/RepositoryResponse.java]: This check does not match criteria:
`If statusCode=200 and no body - return as it is (empty body) keeping the headers and statusCode from repository`.
Currently variable `shouldProcess` is not used anywhere. Shouldn't be here:
```java
if (shouldProcess) {
    repositoryResponse = success(data, headers);
} else {
....
```
?

File: knotx-core/knotx-repository/src/main/java/com/cognifide/knotx/repository/impl/LocalRepository.java [hunk 9]
```diff
@@ -61,10 +63,18 @@ static LocalRepository of(String path, String catalogue, FileSystem fileSystem)
         return fileSystem.openObservable(localFilePath, new OpenOptions())
                 .flatMap(AsyncFile::toObservable)
                 .flatMap(buffer -> RepositoryResponse.success(buffer.toString(), MultiMap.caseInsensitiveMultiMap()).toObservable())
-                .defaultIfEmpty(RepositoryResponse.error("No Template found for path <%s>", repositoryRequest.getPath()))
+                .defaultIfEmpty(RepositoryResponse.success("", MultiMap.caseInsensitiveMultiMap()))
```
Comment (Skejven) [inline knotx-core/knotx-repository/src/main/java/com/cognifide/knotx/repository/impl/LocalRepository.java]: Should be `success` returned when no file of template in local repo found?
Maybe there should be:
`.defaultIfEmpty(RepositoryResponse.error(404, "", MultiMap.caseInsensitiveMultiMap()))` instead?

File: knotx-core/knotx-server/src/main/java/com/cognifide/knotx/server/KnotxServerVerticle.java [hunk 10]
```diff
@@ -71,34 +71,34 @@ public void start(Future<Void> fut) throws IOException, URISyntaxException {
...
+                                        if (repository.shouldProcess()) {
+                                            eventBus.<JsonObject>sendObservable(engineAddress, createEngineRequest(repository, request))
+                                                    .subscribe(
+                                                            result -> {
+                                                                TemplateEngineResponse engineResponse = new TemplateEngineResponse(result.body());
+                                                                if (engineResponse.isSuccess()) {
+                                                                    rewriteHeaders(request, request.headers());
+                                                                    request.response().end(engineResponse.getHtml());
+                                                                } else {
+                                                                    request.response().setStatusCode(HttpResponseStatus.INTERNAL_SERVER_ERROR.code()).end(engineResponse.getReason());
```
Comment (Skejven) [inline knotx-core/knotx-server/src/main/java/com/cognifide/knotx/server/KnotxServerVerticle.java]: In one place we use plain integers as status codes (e.g. `RepositoryResponse`). Here those statuses are taken from `HttpResponseStatus`. Please unify status codes usage (preferably `HttpResponseStatus` ).

File: knotx-core/knotx-api/src/main/java/com/cognifide/knotx/api/RepositoryResponse.java (hunk 7 above)
Comment (arekwolk) [inline knotx-core/knotx-api/src/main/java/com/cognifide/knotx/api/RepositoryResponse.java]: This class doesnt need serialVersionUID, I'm going to remove it

File: knotx-core/knotx-server/src/main/java/com/cognifide/knotx/server/KnotxServerVerticle.java (hunk 10 above)
Comment (arekwolk) [inline knotx-core/knotx-server/src/main/java/com/cognifide/knotx/server/KnotxServerVerticle.java]: Unified in whole app

File: knotx-core/knotx-repository/src/main/java/com/cognifide/knotx/repository/impl/LocalRepository.java (hunk 9 above)
Comment (arekwolk) [inline knotx-core/knotx-repository/src/main/java/com/cognifide/knotx/repository/impl/LocalRepository.java]: You are right. Fixed

File: knotx-core/knotx-api/src/main/java/com/cognifide/knotx/api/RepositoryResponse.java (hunk 8 above)
Comment (arekwolk) [inline knotx-core/knotx-api/src/main/java/com/cognifide/knotx/api/RepositoryResponse.java]: Yes, I missed it. Fixed

**Comment by Skejven** [comment]
Accepted

**Pull request merged**